waiter/src/waiter/scheduler.clj

```diff
@@ -638,3 +642,145 @@
           "WAITER_PASSWORD" (service-id->password-fn service-id)
           "WAITER_SERVICE_ID" service-id
           "WAITER_USERNAME" "waiter"}))
+
+;; Support for tracking service instance launching time stats
+
+(def instance-counts-zero {:requested 0 :started 0})
+
+(defn- make-launch-tracker-entry
+  [service-id]
+  [service-id
+   {:instance-counts instance-counts-zero
+    :known-instance-ids #{}
+    :instance-scheduling-start-times []
+    :service-schedule-timer (metrics/service-timer service-id "launch-overhead" "schedule-time")
+    :service-startup-timer (metrics/service-timer service-id "launch-overhead" "startup-time")
+    :starting-instance-ids []}])
+
+(defn update-launch-trackers
+  "Track metrics on app instances, specifically schedule and startup times."
+  [service-id->launch-tracker new-service-ids removed-service-ids service-id->healthy-instances
+   service-id->unhealthy-instances service-id->instance-counts waiter-schedule-timer]
+  (let [now (t/now)
+        new-tracker-entries (map make-launch-tracker-entry new-service-ids)
+        tracker-entries' (->> service-id->launch-tracker
+                              ;; Remove trackers from deleted services
+                              (remove (comp removed-service-ids key))
+                              ;; Add trackers for newly-discovered services
+                              (concat new-tracker-entries))]
+    ;; Update all trackers...
+    (pc/for-map [[service-id
+                  {:keys [known-instance-ids instance-counts instance-scheduling-start-times
+                          service-schedule-timer service-startup-timer starting-instance-ids]
+                   :as tracker-state}]
+                 tracker-entries']
+      service-id
+      (let [healthy-instances (get service-id->healthy-instances service-id)
+            known-instances' (->> service-id
+                                  (get service-id->unhealthy-instances)
+                                  (into healthy-instances))
+            previously-known-instance? (comp known-instance-ids :id)
+            new-instance-ids (->> known-instances'
+                                  (remove previously-known-instance?)
+                                  (sort-by :started-at)
```

Comment:
Should we sort by `scheduler/instance-comparator` instead?

Reply:
> "The comparison order is: service-id, started-at, and finally id."

We want `:started-at` as the primary sorting key, so that doesn't seem to work.

Reply:
`service-id` will be the same for the two instances, so we will get the primary sorting by `started-at` and ties will be broken by `id` in a deterministic way.

Reply:
Oh! That makes sense now. My bad.